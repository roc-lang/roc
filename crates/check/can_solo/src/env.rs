use std::path::Path;

use bumpalo::Bump;
use roc_module::symbol::ModuleId;
use roc_problem::can::Problem;
use roc_region::all::LineInfo;

/// The canonicalization environment for a particular module.
#[derive(Debug)]
pub struct SoloEnv<'a> {
    pub arena: &'a Bump,

    pub module_path: &'a Path,

    pub solo_home: ModuleId,

    /// Problems we've encountered along the way, which will be reported to the user at the end.
    pub problems: Vec<Problem>,

    pub src: &'a str,

    /// Lazily calculated line info. This data is only needed if the code contains calls to `dbg`,
    /// otherwise we can leave it as `None` and never pay the cost of scanning the source an extra
    /// time.
    pub lazy_line_info: &'a mut Option<LineInfo>,
}

impl<'a> SoloEnv<'a> {
    #[allow(clippy::too_many_arguments)]
    pub fn new(arena: &'a Bump, src: &'a str, module_path: &'a Path) -> SoloEnv<'a> {
        SoloEnv {
            arena,
            module_path,
            solo_home: ModuleId::first_after_builtins(),
            src,
            problems: Vec::new(),
            lazy_line_info: arena.alloc(None),
        }
    }

    pub fn line_info(&mut self) -> &LineInfo {
        if self.lazy_line_info.is_none() {
            *self.lazy_line_info = Some(LineInfo::new(self.src));
        }

        self.lazy_line_info.as_ref().unwrap()
    }
}
