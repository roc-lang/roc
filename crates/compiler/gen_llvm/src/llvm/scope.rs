use inkwell::{
    basic_block::BasicBlock,
    values::{BasicValue, BasicValueEnum, FunctionValue, PhiValue, PointerValue},
};
use roc_collections::ImMap;
use roc_module::symbol::{ModuleId, Symbol};
use roc_mono::{
    ir::{JoinPointId, Param, ProcLayout},
    layout::InLayout,
};

#[derive(Default, Debug, Clone, PartialEq, Eq)]
pub(crate) struct Scope<'a, 'ctx> {
    symbols: ImMap<Symbol, (InLayout<'a>, BasicValueEnum<'ctx>)>,
    top_level_thunks: ImMap<Symbol, (ProcLayout<'a>, FunctionValue<'ctx>)>,
    join_points: ImMap<JoinPointId, (BasicBlock<'ctx>, Vec<JoinPointArg<'ctx>>)>,
}

#[derive(Debug)]
pub(crate) struct JoinPointNotFound;

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) enum JoinPointArg<'ctx> {
    Alloca(PointerValue<'ctx>),
    Phi(PhiValue<'ctx>),
}

impl<'a, 'ctx> Scope<'a, 'ctx> {
    pub fn insert(&mut self, symbol: Symbol, layout: InLayout<'a>, value: BasicValueEnum<'ctx>) {
        self.symbols.insert(symbol, (layout, value));
    }

    pub fn load_symbol(&self, symbol: &Symbol) -> BasicValueEnum<'ctx> {
        match self.symbols.get(symbol) {
            Some((_, ptr)) => *ptr,

            None => panic!("There was no entry for {symbol:?} {symbol} in scope {self:?}"),
        }
    }

    pub fn load_symbol_and_layout(&self, symbol: &Symbol) -> (BasicValueEnum<'ctx>, InLayout<'a>) {
        match self.symbols.get(symbol) {
            Some((layout, ptr)) => (*ptr, *layout),
            None => panic!("There was no entry for {symbol:?} in scope {self:?}"),
        }
    }

    pub fn insert_top_level_thunk(
        &mut self,
        symbol: Symbol,
        layout: ProcLayout<'a>,
        function_value: FunctionValue<'ctx>,
    ) {
        self.top_level_thunks
            .insert(symbol, (layout, function_value));
    }

    pub fn remove(&mut self, symbol: &Symbol) {
        self.symbols.remove(symbol);
    }

    pub fn retain_top_level_thunks_for_module(&mut self, module_id: ModuleId) {
        self.top_level_thunks
            .retain(|s, _| s.module_id() == module_id);
    }

    pub fn insert_join_point(
        &mut self,
        join_point_id: JoinPointId,
        bb: BasicBlock<'ctx>,
        phis: Vec<JoinPointArg<'ctx>>,
    ) {
        self.join_points.insert(join_point_id, (bb, phis));
    }

    pub fn remove_join_point(&mut self, join_point_id: JoinPointId) {
        self.join_points.remove(&join_point_id);
    }

    pub fn get_join_point(
        &self,
        join_point_id: JoinPointId,
    ) -> Option<&(BasicBlock<'ctx>, Vec<JoinPointArg<'ctx>>)> {
        self.join_points.get(&join_point_id)
    }

    pub fn bind_parameters_to_join_point(
        &mut self,
        join_point_id: JoinPointId,
        parameters: impl IntoIterator<Item = &'a Param<'a>>,
    ) -> Result<(), JoinPointNotFound> {
        let ref_join_points = &self
            .join_points
            .get(&join_point_id)
            .ok_or(JoinPointNotFound)?
            .1;

        for (joinpoint_arg, param) in ref_join_points.iter().zip(parameters.into_iter()) {
            match joinpoint_arg {
                crate::llvm::scope::JoinPointArg::Alloca(alloca) => {
                    self.symbols
                        .insert(param.symbol, (param.layout, alloca.as_basic_value_enum()));
                }
                crate::llvm::scope::JoinPointArg::Phi(phi) => {
                    self.symbols
                        .insert(param.symbol, (param.layout, phi.as_basic_value()));
                }
            }
        }

        Ok(())
    }
}
