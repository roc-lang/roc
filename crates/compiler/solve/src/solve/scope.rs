use roc_can::module::ModuleParams;
use roc_module::symbol::Symbol;
use roc_types::subs::Variable;

/// The scope of the solver, as symbols are introduced.
#[derive(Clone, Debug)]
pub struct Scope {
    symbols: Vec<Symbol>,
    variables: Vec<Variable>,
}

impl Scope {
    pub fn new(opt_module_params: Option<ModuleParams>) -> Self {
        match opt_module_params {
            Some(module_params) => {
                let mut symbols = Vec::with_capacity(module_params.destructs.len());
                let mut variables = Vec::with_capacity(module_params.destructs.len());

                for destruct in module_params.destructs {
                    symbols.push(destruct.value.symbol);
                    variables.push(destruct.value.var);
                }

                Self { symbols, variables }
            }
            None => Self {
                symbols: Vec::default(),
                variables: Vec::default(),
            },
        }
    }

    pub fn vars_by_symbol(&self) -> impl Iterator<Item = (Symbol, Variable)> + '_ {
        let it1 = self.symbols.iter().copied();
        let it2 = self.variables.iter().copied();

        it1.zip(it2)
    }

    #[inline(always)]
    pub fn get_var_by_symbol(&self, symbol: &Symbol) -> Option<Variable> {
        self.symbols
            .iter()
            .position(|s| s == symbol)
            .map(|index| self.variables[index])
    }

    #[inline(always)]
    pub fn insert_symbol_var_if_vacant(&mut self, symbol: Symbol, var: Variable) {
        match self.symbols.iter().position(|s| *s == symbol) {
            None => {
                // symbol is not in vars_by_symbol yet; insert it
                self.symbols.push(symbol);
                self.variables.push(var);
            }
            Some(_) => {
                // do nothing
            }
        }
    }
}
