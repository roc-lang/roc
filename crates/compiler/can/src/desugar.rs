#![allow(clippy::manual_map)]

use crate::env::Env;
use crate::scope::Scope;
use roc_module::called_via::CalledVia;
use roc_module::ident::ModuleName;
use roc_parse::ast::Expr::{self, *};
use roc_parse::ast::{Collection, Defs, Pattern, ValueDef};
use roc_region::all::{Loc, Region};

/// Desugars a `dbg expr` expression into a statement block that prints and returns the
/// value produced by `expr`. Essentially:
/// (
///     tmpVar = expr
///     LowLevelDbg (Inspect.to_str tmpVar)
///     tmpVar
/// )
fn desugar_dbg_expr<'a>(
    env: &mut Env<'a>,
    scope: &mut Scope,
    expr: &'a Loc<Expr<'a>>,
    outer_region: Region,
) -> &'a Expr<'a> {
    let region = expr.region;

    // tmpVar = expr
    let ident = env.arena.alloc(scope.gen_unique_symbol_name().to_string());

    let value_def = ValueDef::Body(
        env.arena.alloc(Loc {
            value: Pattern::Identifier { ident },
            region,
        }),
        expr,
    );

    let defs = env.arena.alloc(Defs::default());
    defs.push_value_def(value_def, region, &[], &[]);

    // tmpVar
    let tmp_var = env.arena.alloc(Loc {
        value: Var {
            module_name: "",
            ident,
        },
        region,
    });

    // LowLevelDbg
    let dbg_stmt = env.arena.alloc(Loc {
        value: *desugar_dbg_stmt(env, tmp_var, tmp_var),
        region: outer_region,
    });

    env.arena.alloc(Defs(defs, dbg_stmt))
}

/// Build a desugared `dbg {}` expression to act as a placeholder when the AST
/// is invalid.
pub fn desugar_invalid_dbg_expr<'a>(
    env: &mut Env<'a>,
    scope: &mut Scope,
    outer_region: Region,
) -> &'a Expr<'a> {
    let placeholder_expr = env.arena.alloc(Loc {
        value: Record(Collection::empty()),
        region: outer_region,
    });

    desugar_dbg_expr(env, scope, placeholder_expr, outer_region)
}

/// Desugars a `dbg x` statement into essentially `Inspect.to_str x |> LowLevelDbg`
fn desugar_dbg_stmt<'a>(
    env: &mut Env<'a>,
    condition: &'a Loc<Expr<'a>>,
    continuation: &'a Loc<Expr<'a>>,
) -> &'a Expr<'a> {
    let region = condition.region;

    let inspect_fn = Var {
        module_name: ModuleName::INSPECT,
        ident: "to_str",
    };
    let loc_inspect_fn_var = env.arena.alloc(Loc {
        value: inspect_fn,
        region,
    });
    let inspect_args = &*env.arena.alloc([condition]);

    let dbg_str = env.arena.alloc(Loc {
        value: Apply(loc_inspect_fn_var, inspect_args, CalledVia::Space),
        region,
    });

    let line_col = env.line_info().convert_pos(region.start());

    let dbg_src = env
        .src
        .split_at(region.start().offset as usize)
        .1
        .split_at((region.end().offset - region.start().offset) as usize)
        .0;

    let module_path_str = env.module_path.to_string_lossy();

    // |> LowLevelDbg
    env.arena.alloc(LowLevelDbg(
        env.arena.alloc((
            &*env
                .arena
                .alloc_str(&format!("{}:{}", module_path_str, line_col.line + 1)),
            &*env.arena.alloc_str(dbg_src),
        )),
        dbg_str,
        continuation,
    ))
}
