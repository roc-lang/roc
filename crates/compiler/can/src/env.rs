use std::path::Path;

use crate::procedure::References;
use crate::scope::{ModuleLookup, Scope, SymbolLookup};
use bumpalo::Bump;
use roc_collections::{MutMap, VecSet};
use roc_module::ident::{Ident, ModuleName};
use roc_module::symbol::{IdentIdsByModule, ModuleId, PQModuleName, PackageModuleIds, Symbol};
use roc_problem::can::{Problem, RuntimeError};
use roc_region::all::{LineInfo, Loc, Region};
use roc_types::subs::Variable;

/// The canonicalization environment for a particular module.
pub struct Env<'a> {
    /// The module's path. Opaques and unqualified references to identifiers
    /// are assumed to be relative to this path.
    pub home: ModuleId,

    pub module_path: &'a Path,

    pub dep_idents: &'a IdentIdsByModule,

    pub qualified_module_ids: &'a PackageModuleIds<'a>,

    /// Problems we've encountered along the way, which will be reported to the user at the end.
    pub problems: Vec<Problem>,

    /// Closures
    pub closures: MutMap<Symbol, References>,

    /// current tail-callable symbol
    pub tailcallable_symbol: Option<Symbol>,

    /// Symbols of values/functions which were referenced by qualified lookups.
    pub qualified_value_lookups: VecSet<Symbol>,

    pub top_level_symbols: VecSet<Symbol>,

    pub home_params_record: Option<(Symbol, Variable)>,

    pub arena: &'a Bump,

    pub opt_shorthand: Option<&'a str>,

    pub fx_mode: FxMode,

    pub src: &'a str,

    /// Lazily calculated line info. This data is only needed if the code contains calls to `dbg`,
    /// otherwise we can leave it as `None` and never pay the cost of scanning the source an extra
    /// time.
    line_info: &'a mut Option<LineInfo>,
}

impl<'a> Env<'a> {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        arena: &'a Bump,
        src: &'a str,
        home: ModuleId,
        module_path: &'a Path,
        dep_idents: &'a IdentIdsByModule,
        qualified_module_ids: &'a PackageModuleIds<'a>,
        opt_shorthand: Option<&'a str>,
        fx_mode: FxMode,
    ) -> Env<'a> {
        Env {
            arena,
            src,
            home,
            module_path,
            dep_idents,
            qualified_module_ids,
            problems: Vec::new(),
            closures: MutMap::default(),
            qualified_value_lookups: VecSet::default(),
            tailcallable_symbol: None,
            top_level_symbols: VecSet::default(),
            home_params_record: None,
            opt_shorthand,
            line_info: arena.alloc(None),
            fx_mode,
        }
    }

    pub fn qualified_lookup(
        &mut self,
        scope: &Scope,
        module_name_str: &str,
        ident: &str,
        region: Region,
    ) -> Result<SymbolLookup, RuntimeError> {
        debug_assert!(
            !module_name_str.is_empty(),
            "Called env.qualified_lookup with an unqualified ident: {ident:?}"
        );

        let module_name = ModuleName::from(module_name_str);

        match scope.modules.lookup(&module_name) {
            Some(lookedup_module) => {
                self.qualified_lookup_help(scope, lookedup_module, ident, region)
            }
            None => Err(RuntimeError::ModuleNotImported {
                module_name: module_name.clone(),
                imported_modules: scope
                    .modules
                    .available_names()
                    .map(|string| string.as_ref().into())
                    .collect(),
                region,
                module_exists: self
                    .qualified_module_ids
                    .get_id(&PQModuleName::Unqualified(module_name))
                    .is_some(),
            }),
        }
    }

    pub fn qualified_lookup_with_module_id(
        &mut self,
        scope: &Scope,
        module_id: ModuleId,
        ident: &str,
        region: Region,
    ) -> Result<SymbolLookup, RuntimeError> {
        if let Some(module) = scope.modules.lookup_by_id(&module_id) {
            self.qualified_lookup_help(scope, module, ident, region)
        } else {
            Err(self.module_exists_but_not_imported(scope, module_id, region))
        }
    }

    /// Returns Err if the symbol resolved, but it was not exposed by the given module
    fn qualified_lookup_help(
        &mut self,
        scope: &Scope,
        module: ModuleLookup,
        ident: &str,
        region: Region,
    ) -> Result<SymbolLookup, RuntimeError> {
        let is_type_name = ident.starts_with(|c: char| c.is_uppercase());

        // You can do qualified lookups on your own module, e.g.
        // if I'm in the Foo module, I can do a `Foo.bar` lookup.
        if module.id == self.home {
            match scope.locals.ident_ids.get_id(ident) {
                Some(ident_id) => {
                    let symbol = Symbol::new(module.id, ident_id);

                    if !is_type_name {
                        self.qualified_value_lookups.insert(symbol);
                    }

                    Ok(SymbolLookup::no_params(symbol))
                }
                None => {
                    let error = RuntimeError::LookupNotInScope {
                        loc_name: Loc {
                            value: Ident::from(ident),
                            region,
                        },
                        suggestion_options: scope
                            .locals
                            .ident_ids
                            .ident_strs()
                            .map(|(_, string)| string.into())
                            .collect(),
                        underscored_suggestion_region: None,
                    };
                    Err(error)
                }
            }
        } else {
            match self.dep_idents.get(&module.id) {
                Some(exposed_ids) => match exposed_ids.get_id(ident) {
                    Some(ident_id) => {
                        let symbol = Symbol::new(module.id, ident_id);

                        if !is_type_name {
                            self.qualified_value_lookups.insert(symbol);
                        }

                        Ok(module.into_symbol(symbol))
                    }
                    None => Err(RuntimeError::ValueNotExposed {
                        module_name: self
                            .qualified_module_ids
                            .get_name(module.id)
                            .expect("Module ID known, but not in the module IDs somehow")
                            .as_inner()
                            .clone(),
                        ident: Ident::from(ident),
                        region,
                        exposed_values: exposed_ids.exposed_values(),
                    }),
                },
                _ => Err(self.module_exists_but_not_imported(scope, module.id, region)),
            }
        }
    }

    fn module_exists_but_not_imported(
        &self,
        scope: &Scope,
        module_id: ModuleId,
        region: Region,
    ) -> RuntimeError {
        RuntimeError::ModuleNotImported {
            module_name: self
                .qualified_module_ids
                .get_name(module_id)
                .expect("Module ID known, but not in the module IDs somehow")
                .as_inner()
                .clone(),
            imported_modules: scope
                .modules
                .available_names()
                .map(|string| string.as_ref().into())
                .collect(),
            region,
            module_exists: true,
        }
    }

    pub fn problem(&mut self, problem: Problem) {
        self.problems.push(problem)
    }

    pub fn line_info(&mut self) -> &LineInfo {
        if self.line_info.is_none() {
            *self.line_info = Some(LineInfo::new(self.src));
        }
        self.line_info.as_ref().unwrap()
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum FxMode {
    PurityInference,
    Task,
}
