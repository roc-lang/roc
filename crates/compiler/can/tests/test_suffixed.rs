#[macro_use]
extern crate indoc;

#[cfg(test)]
mod suffixed_tests {
    use bumpalo::Bump;
    use insta::assert_snapshot;
    use roc_can::desugar::desugar_defs_node_values;
    use roc_can::env::{Env, FxMode};
    use roc_can::scope::Scope;
    use roc_module::symbol::{IdentIds, ModuleIds, PackageModuleIds};
    use roc_parse::test_helpers::parse_defs_with;
    use std::path::Path;

    macro_rules! run_test {
        ($src:expr) => {{
            let arena = &Bump::new();
            let home = ModuleIds::default().get_or_insert(&"Test".into());

            let mut scope = Scope::new(
                home,
                "TestPath".into(),
                IdentIds::default(),
                Default::default(),
            );

            let dep_idents = IdentIds::exposed_builtins(0);
            let qualified_module_ids = PackageModuleIds::default();
            let mut env = Env::new(
                arena,
                $src,
                home,
                Path::new("test.roc"),
                &dep_idents,
                &qualified_module_ids,
                None,
                FxMode::Task,
            );

            let mut defs = parse_defs_with(arena, indoc!($src)).unwrap();
            desugar_defs_node_values(&mut env, &mut scope, &mut defs, true);

            let snapshot = format!("{:#?}", &defs);
            println!("{}", snapshot);
            assert_snapshot!(snapshot);
        }};
    }

    /**
     * This example tests a suffixed statement, followed
     * by a Body with an empty record pattern.
     *
     * The def final expression is explicitly provided.
     */
    #[test]
    fn multi_defs_stmts() {
        run_test!(
            r#"
            main =
                line! "Ahoy"
                {} = "There" |> Stdout.line!

                Task.ok {}
            "#
        );
    }

    /**
     * The most simple suffixed example. A single statement
     * without arguments and a final expression.
     */
    #[test]

    fn basic() {
        run_test!(
            r#"
        main =
            foo!

            ok {}
        "#
        );
    }

    /**
     * A single suffixed statement with arguments applied.
     * Note there is no final expression.
     */
    #[test]
    fn last_suffixed_single() {
        run_test!(
            r#"
            main = foo! "bar" {} "baz"
            "#
        );
    }

    /**
     * Multiple suffixed statements with no
     * arguments, and no final expression.
     */
    #[test]
    fn last_suffixed_multiple() {
        run_test!(
            r#"
            main =
                foo!
                bar!
                baz!
            "#
        );
    }

    /**
     * A definition with a closure that contains a Defs node, which also
     * contains a suffixed binops statement.
     */
    #[test]
    fn closure_simple() {
        run_test!(
            r#"
            main =
                x = \msg ->
                    msg |> line!
                    ok {}

                x "hi"
            "#
        );
    }

    /**
     * Example of unwrapping a pipline statement
     *
     * Note pipelines are desugared into Apply functions,
     * however this also tests the parser.
     *
     */
    #[test]
    fn simple_pizza() {
        run_test!(
            r#"
            main =
                "hello"
                |> Str.concat "world"
                |> line!

                Task.ok {}
            "#
        );
    }

    /**
     * Example with a parens suffixed sub-expression
     * in the function part of an Apply.
     *
     * Note how the parens unwraps into an intermediate answer #!0_arg instead of
     * unwrapping the def `do`.
     *
     */
    #[test]
    fn body_parens_apply() {
        run_test!(
            r#"
            main =
                do = (sayMultiple!) "hi"
                do
            "#
        );
    }

    /**
     * Example of unwrapping mixed Body defs with
     * Var's of both single and multiple suffixes
     */
    #[test]
    fn var_suffixes() {
        run_test!(
            r#"
            main =
                a = foo!
                b = bar!!
                baz a b
            "#
        );
    }

    /**
     * Example with a multiple suffixed Var
     *
     * Note it unwraps into an intermediate answer `#!0_arg`
     *
     */
    #[test]
    fn multiple_suffix() {
        run_test!(
            r#"
            main =
                foo!!
                bar
            "#
        );
    }

    /**
     * A suffixed expression in the function part of the Apply
     */
    #[test]
    fn apply_function_suffixed() {
        run_test!(
            r#"
            main =
                x = (foo! "bar") "hello"
                baz x
            "#
        );
    }

    /**
     * A suffixed expression in an Apply argument position.
     */
    #[test]
    fn apply_argument_suffixed() {
        run_test!(
            r#"
            main =
                x = bar (foo! "hello")
                baz x
            "#
        );
    }

    /**
     * Example where the suffixed def is not the first def
     */
    #[test]
    fn multiple_def_first_suffixed() {
        run_test!(
            r#"
            main =
                msg = "hello"
                x = foo! msg
                bar x
            "#
        );
    }

    /**
     * Annotated defs and a suffixed expression
     * with annotations inside a closure
     */

    #[test]
    fn closure_with_annotations() {
        run_test!(
            r#"
            main =
                x : Str -> Task _ _
                x = \msg ->

                    y : Task {} _
                    y = line! msg
                    y

                x "foo"
            "#
        );
    }

    /**
     * Nested suffixed expressions
     */
    #[test]
    fn nested_simple() {
        run_test!(
            r#"
            run = line! (nextMsg!)
            "#
        );
    }

    /**
     * Nested suffixed expressions
     */
    #[test]
    fn nested_complex() {
        run_test!(
            r#"
            main =
                z = foo! (bar! baz) (blah stuff)
                doSomething z
            "#
        );
    }

    /**
     * A closure that contains a Defs node
     */
    #[test]
    fn closure_with_defs() {
        run_test!(
            r#"
            main =

                foo : Str, {}, Str -> Task {} I32
                foo = \a, _, b ->
                    line! a
                    line! b

                    Task.ok {}

                foo "bar" {} "baz"
            "#
        );
    }

    /**
     * Test when the suffixed def being unwrapped is not the first or last
     */
    #[test]
    fn defs_suffixed_middle() {
        run_test!(
            r#"
            main =
                a = "Foo"
                Stdout.line! a

                printBar!

            printBar =
                b = "Bar"
                Stdout.line b
            "#
        );
    }

    /**
     * A simple if-then-else statement which is split
     */
    #[test]
    fn if_simple() {
        run_test!(
            r#"
            main =
                isTrue = Task.ok Bool.true
                isFalse = Task.ok Bool.false

                if isFalse! then
                    line "fail"
                else if isTrue! then
                    line "success"
                else
                    line "fail"
            "#
        );
    }

    /**
     * A more complex example including the use of nested Defs nodes
     */
    #[test]
    fn if_complex() {
        run_test!(
            r#"
            main =
                isTrue = Task.ok Bool.true
                isFalsey = \x -> Task.ok x
                msg : Task {} I32
                msg =
                    if !(isTrue!) then
                        line! "fail"
                        err 1
                    else if (isFalsey! Bool.false) then
                        line! "nope"
                        ok {}
                    else
                        line! "success"

                msg
            "#
        );
    }

    /**
     * Unwrap a trailing binops
     */
    #[test]
    fn trailing_binops() {
        run_test!(
            r#"
            copy = \a,b ->
                line! "FOO"

                CMD.new "cp"
                |> mapErr! ERR
            "#
        );
    }

    /**
     * Unwrap a when expression
     */
    #[test]
    fn when_simple() {
        run_test!(
            r#"
            list =
                when getList! is
                    [] -> "empty"
                    _ -> "non-empty"
            "#
        );
    }

    /**
     * Unwrap a when expression
     */
    #[test]
    fn when_branches() {
        run_test!(
            r#"
            list =
                when getList! is
                    [] ->
                        line! "foo"
                        line! "bar"
                    _ ->
                        ok {}
            "#
        );
    }

    #[test]
    fn trailing_suffix_inside_when() {
        run_test!(
            r#"
            main =
                result = Stdin.line!

                when result is
                    End ->
                        Task.ok {}

                    Input name ->
                        Stdout.line! "Hello, $(name)"
            "#
        );
    }

    #[test]
    fn dbg_simple() {
        run_test!(
            r#"
            main =
                foo = getFoo!
                dbg foo
                bar! foo
            "#
        );
    }

    #[test]
    fn dbg_expr() {
        run_test!(
            r#"
            main =
                dbg (dbg (1 + 1))
            "#
        );
    }

    #[test]
    fn pizza_dbg() {
        run_test!(
            r#"
            main =
                1
                |> dbg
                |> Num.add 2
                |> dbg
            "#
        )
    }

    #[test]
    fn apply_argument_single() {
        run_test!(
            r#"
            main =
                c = b a!
                c
            "#
        );
    }

    #[test]
    fn apply_argument_multiple() {
        run_test!(
            r#"
            main =
                c = b a! x!
                c
            "#
        );
    }

    #[test]
    fn bang_in_pipe_root() {
        run_test!(
            r#"
            main =
                c = a! |> b
                c
            "#
        );
    }

    #[test]
    fn expect_then_bang() {
        run_test!(
            r#"
            main =
                expect 1 == 2
                x!
            "#
        );
    }

    #[test]
    fn deep_when() {
        run_test!(
            r#"
            main =
                when a is
                    0 ->
                        when b is
                            1 ->
                                c!
            "#
        );
    }

    #[test]
    fn deps_final_expr() {
        run_test!(
            r#"
            main =
                when x is
                    A ->
                        y = 42

                        if a then
                            b!
                        else
                            c!
                    B ->
                        d!
            "#
        );
    }

    #[test]
    fn dbg_stmt_arg() {
        run_test!(
            r#"
            main =
                dbg a!

                b
            "#
        )
    }

    #[test]
    fn last_stmt_not_top_level_suffixed() {
        run_test!(
            r#"
            main =
                x = 42
                a b!
            "#
        );
    }

    #[test]
    fn nested_defs() {
        run_test!(
            r##"
            main =
                x =
                    a = b!
                    c! a

                x
            "##
        );
    }

    #[test]
    fn type_annotation() {
        run_test!(
            r##"
            f = \x ->
                r : A
                r = x!
                Task.ok r
            "##
        );
    }

    #[test]
    fn issue_7081() {
        run_test!(
            r##"
            inc = \i ->
                if i > 2 then
                    Err MaxReached
                else
                    Ok (i + 1)

            expect
                run = \i ->
                    newi =
                        i
                        |> inc?
                        |> inc?
                    Ok newi
                result = run 0
                result == Ok 2

            main =
                Stdout.line! "Hello world"
            "##
        );
    }

    #[test]
    fn issue_7103() {
        run_test!(
            r##"
            run : Task {} _
            run = line! "foo"

            main = run
            "##
        );
    }
}

#[cfg(test)]
mod test_suffixed_helpers {

    use roc_can::suffixed::is_matching_intermediate_answer;
    use roc_parse::ast::Expr;
    use roc_parse::ast::Pattern;
    use roc_region::all::Loc;

    #[test]
    fn test_matching_answer() {
        let loc_pat = Loc::at_zero(Pattern::Identifier { ident: "#!0_arg" });
        let loc_new = Loc::at_zero(Expr::Var {
            module_name: "",
            ident: "#!0_arg",
        });

        std::assert!(is_matching_intermediate_answer(&loc_pat, &loc_new));
    }
}
