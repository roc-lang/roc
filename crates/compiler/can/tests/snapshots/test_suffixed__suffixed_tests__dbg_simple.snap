---
source: crates/compiler/can/tests/test_suffixed.rs
expression: snapshot
snapshot_kind: text
---
Defs {
    tags: [
        EitherIndex(2147483648),
    ],
    regions: [
        @0-49,
    ],
    space_before: [
        Slice<roc_parse::ast::CommentOrNewline> { start: 0, length: 0 },
    ],
    space_after: [
        Slice<roc_parse::ast::CommentOrNewline> { start: 0, length: 1 },
    ],
    spaces: [
        Newline,
    ],
    type_defs: [],
    value_defs: [
        Body(
            @0-4 Identifier {
                ident: "main",
            },
            @17-23 Apply(
                @17-23 Var {
                    module_name: "Task",
                    ident: "await",
                },
                [
                    @17-23 Var {
                        module_name: "",
                        ident: "getFoo",
                    },
                    @17-23 Closure(
                        [
                            @11-14 Identifier {
                                ident: "foo",
                            },
                        ],
                        @29-36 LowLevelDbg(
                            (
                                "test.roc:3",
                                "   ",
                            ),
                            @33-36 Apply(
                                @33-36 Var {
                                    module_name: "Inspect",
                                    ident: "toStr",
                                },
                                [
                                    @33-36 Var {
                                        module_name: "",
                                        ident: "foo",
                                    },
                                ],
                                Space,
                            ),
                            @41-49 Apply(
                                @41-49 Var {
                                    module_name: "",
                                    ident: "bar",
                                },
                                [
                                    @46-49 Var {
                                        module_name: "",
                                        ident: "foo",
                                    },
                                ],
                                Space,
                            ),
                        ),
                    ),
                ],
                BangSuffix,
            ),
        ),
    ],
}
